### ISSUE TITLE
Type instability in `DataLoader`

### ISSUE THREAD

**Issue by Vilin97** [JuliaML/MLUtils.jl#168]
Is `DataLoader` type-unstable? Seems so. Does it have to be type-unstable?

```julia
using MLUtils
function test_data_loader(data_loader)
    for (x,y) in data_loader
        continue
    end
end
X = rand(Float32, 2, 10)
Y = rand(Float32, 2, 10)
data_loader = MLUtils.DataLoader((data=X, label=Y))
@code_warntype test_data_loader(data_loader) # lots of red
```
If it is type-unstable and it's the intended behavior, it would be great to add a sentence `This function is type-unstable` to the docstring.

**Comment by ToucheSir** [comment]
I don't think it would be impossible to make `iterate(::DataLoader)` type stable, but it would be tricky since there are so many conditionally-applied wrappers here. @lorenzoh and @CarloLucibello thoughts? Do you recall why we moved away from creating the wrappers in the DataLoader constructor?
